### PR TITLE
Build Fix 2

### PR DESCRIPTION
**Pull request by ryanmrichard**
## Description
This PR is meant to replace  (now closed) PR #513.  The overall goal is to build and install Psi4 in a manner that allows it to be included as a dependency in other packages.

## Todos
- [x] Get tests to pass

Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] Supports externally built Pybind11
    - [x] Required pulling in PR #557 (*i.e.* this PR includes that one)
  - [x] No more `LD_PRELOAD` required for MKL
  - [x] Factored plugin copy/pasted code so that the next item need only be changed in one place
  - [x] Plugins now respect `CMAKE_PREFIX_PATH` and do not clobber it
* **User-Facing for Release Notes**
  - [x] `find_package(psi4 CONFIG REQUIRED)` works and respects externally built Pybind11

## Status
- [x]  Ready to go

### CODE REVIEW

File: cmake/math/MathLibs.cmake [hunk 1]
```diff
@@ -145,19 +145,8 @@ else()
...
-set(MKL_BLAS_LIBS3 ${_scalapack_lib} ${_compiler_mkl_interface}${_lib_suffix} ${_thread_lib} mkl_core ${_blacs_lib}         pthread m)
-# ancient MKL BLAS
-set(MKL_BLAS_LIBS4 mkl guide m)
-
-set(MKL_LAPACK_LIBS mkl_lapack95${_lib_suffix} ${_compiler_mkl_interface}${_lib_suffix})
-
-# older MKL LAPACK
-set(MKL_LAPACK_LIBS2 mkl_lapack)
+set(MKL_BLAS_LIBS mkl_rt pthread m dl)
+set(MKL_LAPACK_LIBS mkl_rt pthread m dl)
```
Comment (loriab) [inline cmake/math/MathLibs.cmake]: (1) Why the `dl`?
(2) Why repeat the sealer libs for blas and lapack when they're returned as lapackblas anyway?
(3) Why kill off the ability to return a non-mkl_rt library list by removing all the MKL_BLAS_LIBS{N} signatures? Why not just insert your new one as MKL_BLAS_LIBS and increment all the others by one?
(4) Would be nice to be able to bypass the mkl_rt option with a cmakevar.

File: external/upstream/libefp/CMakeLists.txt [hunk 2]
```diff
@@ -9,7 +9,7 @@ else()
     message(STATUS "Suitable libefp could not be located, ${Magenta}Building libefp${ColourReset} instead.")
     ExternalProject_Add(libefp_external
         GIT_REPOSITORY https://github.com/psi4/libefp
-        GIT_TAG config3
+        GIT_TAG e4e1525858532153ae24ec8ffe59d9d57f221181
```
Comment (loriab) [inline external/upstream/libefp/CMakeLists.txt]: Why? master's pretty far behind.

File: cmake/psi4OptionsTools.cmake [hunk 3]
```diff
@@ -200,4 +200,29 @@ if(${ENABLE_${PLUGIN_NAME}})
...
+    COMMAND ${CMAKE_COMMAND} -E chdir "${CCBD}/${PLUG}" ${CMAKE_MAKE_PROGRAM}
+    COMMAND ${CMAKE_COMMAND} -E create_symlink ${CCBD}/${PLUG}/input.dat ${CCSD}/input.dat
+    COMMAND ${CMAKE_COMMAND} -E create_symlink "${PLUG}/${PLUG}.so" "${PLUG}.so"
+    COMMAND ${CMAKE_COMMAND} -E create_symlink "${PLUG}/__init__.py" "__init__.py"
+    COMMAND ${CMAKE_COMMAND} -E create_symlink "${PLUG}/pymodule.py" "pymodule.py"
+    COMMENT "Build ${PLUG} example plugin"
+    VERBATIM)
+include(TestingMacros)
+add_regression_test(${PLUG} "${test_names}")
+endmacro(add_skeleton_plugin PLUG EXTRA_FLAGS)
```
Comment (loriab) [inline cmake/psi4OptionsTools.cmake]: Ok, nice. FYI, CMake itself has deprecated the crazy repeat-loop-opening-args-in-closing-and-else, so you needn't feel obliged to carry on with it.

**Comment by ryanmrichard** [comment]
1.  Intel says so.  No idea why, I am just following directions.
2. I believe something was still using the variables.  Either way this
mimics the behavior of the original script.
3.  That behavior makes no sense with modern Psi4, i.e. it won't work why
have it?
4.  Again bypass it and do what?  Unless you plan on going back to an
executable this is how to link to MKL.
5.  There's lapack problems otherwise.
6. Thanks, I usually don't put the end stuff, but I was following
established syntax in the file.


Sent from my Android (without spellcheck).  Please forgive the brevity.

On Jan 15, 2017 2:54 PM, "Lori A. Burns" <notifications@github.com> wrote:

*@loriab* commented on this pull request.
------------------------------

In cmake/math/MathLibs.cmake
<https://github.com/psi4/psi4/pull/563#pullrequestreview-16721749>:

> @@ -145,19 +145,8 @@ else()
     set(_blacs_lib)
 endif()

-# miro: for MKL 10.0.1.014
-set(MKL_BLAS_LIBS ${_scalapack_lib}
${_compiler_mkl_interface}${_lib_suffix} ${_thread_lib} mkl_core
mkl_def mkl_mc ${_blacs_lib} guide pthread m)
-#  try this MKL BLAS combination with SGI MPT
-set(MKL_BLAS_LIBS2 ${_scalapack_lib}
${_compiler_mkl_interface}${_lib_suffix} ${_thread_lib} mkl_core
${_blacs_lib}   guide pthread m)
-# newer MKL BLAS versions do not have libguide
-set(MKL_BLAS_LIBS3 ${_scalapack_lib}
${_compiler_mkl_interface}${_lib_suffix} ${_thread_lib} mkl_core
${_blacs_lib}         pthread m)
-# ancient MKL BLAS
-set(MKL_BLAS_LIBS4 mkl guide m)
-
-set(MKL_LAPACK_LIBS mkl_lapack95${_lib_suffix}
${_compiler_mkl_interface}${_lib_suffix})
-
-# older MKL LAPACK
-set(MKL_LAPACK_LIBS2 mkl_lapack)
+set(MKL_BLAS_LIBS mkl_rt pthread m dl)
+set(MKL_LAPACK_LIBS mkl_rt pthread m dl)

(1) Why the dl?
(2) Why repeat the sealer libs for blas and lapack when they're returned as
lapackblas anyway?
(3) Why kill off the ability to return a non-mkl_rt library list by
removing all the MKL_BLAS_LIBS{N} signatures? Why not just insert your new
one as MKL_BLAS_LIBS and increment all the others by one?
(4) Would be nice to be able to bypass the mkl_rt option with a cmakevar.
------------------------------

In external/upstream/libefp/CMakeLists.txt
<https://github.com/psi4/psi4/pull/563#pullrequestreview-16721749>:

> @@ -9,7 +9,7 @@ else()
     message(STATUS "Suitable libefp could not be located,
${Magenta}Building libefp${ColourReset} instead.")
     ExternalProject_Add(libefp_external
         GIT_REPOSITORY https://github.com/psi4/libefp
-        GIT_TAG config3
+        GIT_TAG e4e1525858532153ae24ec8ffe59d9d57f221181

Why? master's pretty far behind.
------------------------------

In cmake/psi4OptionsTools.cmake
<https://github.com/psi4/psi4/pull/563#pullrequestreview-16721749>:

> +add_custom_target(plugin_${PLUG}
+    ALL
+    DEPENDS psi4-core
+    COMMAND ${CMAKE_COMMAND} -E remove_directory ${CCBD}/${PLUG}
+    COMMAND ${PSIEXE} --plugin-name ${PLUG} ${EXTRA_FLAGS}
+    COMMAND ${CMAKE_COMMAND} -E chdir "${CCBD}/${PLUG}" cmake -C
${STAGED_INSTALL_PREFIX}/share/cmake/psi4/psi4PluginCache.cmake
"-DCMAKE_PREFIX_PATH=${DIR_2_PASS}" .
+    COMMAND ${CMAKE_COMMAND} -E chdir "${CCBD}/${PLUG}" ${CMAKE_MAKE_PROGRAM}
+    COMMAND ${CMAKE_COMMAND} -E create_symlink
${CCBD}/${PLUG}/input.dat ${CCSD}/input.dat
+    COMMAND ${CMAKE_COMMAND} -E create_symlink "${PLUG}/${PLUG}.so"
"${PLUG}.so"
+    COMMAND ${CMAKE_COMMAND} -E create_symlink "${PLUG}/__init__.py"
"__init__.py"
+    COMMAND ${CMAKE_COMMAND} -E create_symlink "${PLUG}/pymodule.py"
"pymodule.py"
+    COMMENT "Build ${PLUG} example plugin"
+    VERBATIM)
+include(TestingMacros)
+add_regression_test(${PLUG} "${test_names}")
+endmacro(add_skeleton_plugin PLUG EXTRA_FLAGS)

Ok, nice. FYI, CMake itself has deprecated the crazy
repeat-loop-opening-args-in-closing-and-else, so you needn't feel obliged
to carry on with it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/psi4/psi4/pull/563#pullrequestreview-16721749>, or mute
the thread

**Comment by ryanmrichard** [comment]
@psi4/admins this is ready to go barring any requested changes (failing test case is a time out).

**Comment by loriab** [comment]
Thanks. It honestly _isn't_ my intention to scupper all your PRs, @ryanmrichard. It's just that the `mkl_rt.so` or nothing approach doesn't work with the binary build (and that libefp ref is very far behind). I've been working on a replacement scheme that gets the valuable mkl_rt for most and works with libefp and CheMPS2, too.

If you wanted to do a separate or modified PR with all the plugin stuff, that'd be great. Also, it looks like the pybind 2.0 commit is still caught in this PR, so may need a rebase.

Sorry for the trouble.

**Comment by ryanmrichard** [comment]
So if I'm understanding you correctly basically the only thing you want from this PR is the `CMakeLists.txt` factorization and the (then) one-line change of ensuring `CMAKE_PREFIX_PATH` is passed to the plugins?

**Comment by loriab** [comment]
Yes, please.

**Comment by ryanmrichard** [comment]
Superseded by #582

**Pull request closed without merging**